### PR TITLE
[mono][debugger] Support symbol server on mobile devices

### PR DESCRIPTION
**Pull request by thaystg**
I need these informations to correct create the URL to download symbols from symbol server.

I created a new message to avoid passing all the assembly content over the protocol and read it using PEReader which would be easier, but it would spend more time and more memory on client side also.

Related to PRs on debugger-libs and debugger-vs repo:
https://github.com/xamarin/debugger-vs/pull/315
https://github.com/mono/debugger-libs/pull/382

https://user-images.githubusercontent.com/4503299/221028994-a9e5c64b-ae8f-4c85-865e-8d3631d775db.mp4

### CODE REVIEW

File: src/mono/mono/component/debugger-protocol.h [hunk 1]
```diff
@@ -147,8 +147,8 @@ typedef enum {
 	MDBGPROT_CMD_ASSEMBLY_GET_METHOD_FROM_TOKEN = 12,
 	MDBGPROT_CMD_ASSEMBLY_HAS_DEBUG_INFO = 13,
 	MDBGPROT_CMD_ASSEMBLY_GET_CATTRS = 14,
-	MDBGPROT_CMD_ASSEMBLY_GET_CUSTOM_ATTRIBUTES = 15,
-	MDBGPROT_CMD_ASSEMBLY_GET_PEIMAGE_ADDRESS = 16,
+	MDBGPROT_CMD_ASSEMBLY_GET_PEIMAGE_ADDRESS = 15,
```
Comment (lambdageek) [inline src/mono/mono/component/debugger-protocol.h:150]: What happened to GET_CUSTOM_ATTRIBUTES?

Are we allowed to reuse command indexes?

File: src/mono/mono/metadata/debug-mono-ppdb.c [hunk 2]
```diff
@@ -87,6 +87,15 @@ get_pe_debug_info (MonoImage *image, guint8 *out_guid, gint32 *out_age, gint32 *
 		debug_dir.address         = read32(data + 20);
 		debug_dir.pointer         = read32(data + 24);
 
+		if (pdb_checksum_hash_type && pdb_checksum && debug_dir.type == DEBUG_DIR_PDB_CHECKSUM)
+		{
+			data  = (guint8 *) (image->raw_data + debug_dir.pointer);
+			char* alg_name = (char*)data;
+			guint8*	checksum = (guint8 *) (data + strlen(alg_name)+ 1);
```
Comment (lambdageek) [inline src/mono/mono/metadata/debug-mono-ppdb.c:94]: Should we have a bounds-check here?

File: src/mono/mono/metadata/debug-mono-ppdb.c [hunk 3]
```diff
@@ -96,6 +105,8 @@ get_pe_debug_info (MonoImage *image, guint8 *out_guid, gint32 *out_age, gint32 *
 			if (dir.signature == 0x53445352) {
 				memcpy (out_guid, data + 4, 16);
 				*out_age = read32(data + 20);
+				if (pdb_path)
+					*pdb_path = (char*) data + 24;
```
Comment (lambdageek) [inline src/mono/mono/metadata/debug-mono-ppdb.c:109]: Also here - should we try to find the end of the string and make sure it doesn't go past the end of the debug info

**Review by lambdageek — APPROVED** [review]
LGTM.  But we should probably check the bounds of the PDB data.  And also I'm not sure if it's ok to reuse a command number even if you bump the protocol version.

File: src/mono/mono/metadata/debug-mono-ppdb.c (hunk 2 above)
Comment (thaystg) [reply src/mono/mono/metadata/debug-mono-ppdb.c:94]: I'm not sure, this data will be used on buffer_add_string which will access until find a \0, right?

File: src/mono/mono/component/debugger-protocol.h (hunk 1 above)
Comment (thaystg) [reply src/mono/mono/component/debugger-protocol.h]: Wrong merge, I was testing it on .net7. Fixed thanks.

File: src/mono/mono/metadata/debug-mono-ppdb.c (hunk 2 above)
Comment (thaystg) [reply src/mono/mono/metadata/debug-mono-ppdb.c:94]: Do you mean a case where there isn't a \0 and buffer_add_string will read it and pass the end of the debug info?
I'm not sure if it's possible but I can check if you think we should :)

**Review by lewing — APPROVED** [review]

File: src/mono/mono/metadata/debug-mono-ppdb.h [hunk 4]
```diff
@@ -74,4 +74,7 @@ mono_ppdb_is_embedded (MonoPPDBFile *ppdb);
 MONO_COMPONENT_API MonoPPDBFile*
 mono_create_ppdb_file (MonoImage *ppdb_image, gboolean is_embedded_ppdb);
 
+MONO_COMPONENT_API gboolean
+get_pe_debug_info_full (MonoImage *image, guint8 *out_guid, gint32 *out_age, gint32 *out_timestamp, guint8 **ppdb_data,
```
Comment (akoeplinger) [inline src/mono/mono/metadata/debug-mono-ppdb.h:78]: does this need to be MONO_COMPONENT_API ?

Comment (thaystg) [reply src/mono/mono/metadata/debug-mono-ppdb.h:78]: Thanks! Addressed in this PR: https://github.com/dotnet/runtime/pull/82587/

**Pull request merged**